feat: Allow client type templating

msfstef opened the pull request:
Addresses https://github.com/electric-sql/electric/issues/1466

Work on this was already done for `useShape` by @kevin-dp - this extends it to all of the client APIs.

KyleAMathews reviewed (approved):
Looks great! Excited to have this. Perhaps add typing to an example or two to demo this? Also is there anywhere in the docs we could/should mention this?

kevin-dp commented on packages/typescript-client/src/types.ts:
```diff
@@ -25,7 +27,7 @@ export type ChangeMessage<T> = {
 
 // Define the type for a record
-export type Message<T extends Value = { [key: string]: Value }> =
+export type Message<T extends Row = { [key: string]: Value }> =
```
Can simplify to `export type Message <T extends Row = Row>`

kevin-dp commented on packages/typescript-client/src/types.ts:
```diff
@@ -25,7 +27,7 @@ export type ChangeMessage<T> = {
-export type Message<T extends Value = { [key: string]: Value }> =
+export type Message<T extends Row = { [key: string]: Value }> =
   | ControlMessage
   | ChangeMessage<T>
```
We may also want to add the subclass requirement on `export type ChangeMessage<T extends Row>`

kevin-dp commented on packages/typescript-client/src/types.ts:
```diff
@@ -104,7 +106,7 @@ export type ColumnInfo =
 export type Schema = { [key: string]: ColumnInfo }
 
-export type TypedMessages<T extends Value = { [key: string]: Value }> = {
+export type TypedMessages<T extends Row = { [key: string]: Value }> = {
```
`T extends Row = Row`

kevin-dp commented on packages/typescript-client/src/helpers.ts:
````diff
@@ -17,7 +17,7 @@ import { ChangeMessage, ControlMessage, Message, Value } from './types'
  * ```
  */
-export function isChangeMessage<T extends Value = { [key: string]: Value }>(
+export function isChangeMessage<T extends Row = { [key: string]: Value }>(
````
`T extends Row = Row`

kevin-dp commented on packages/typescript-client/src/helpers.ts:
````diff
@@ -40,7 +40,7 @@ export function isChangeMessage<T extends Value = { [key: string]: Value }>(
  * ```
  */
-export function isControlMessage<T extends Value = { [key: string]: Value }>(
+export function isControlMessage<T extends Row = { [key: string]: Value }>(
````
`T extends Row = Row`

kevin-dp commented on packages/typescript-client/test/integration.test.ts:
```diff
@@ -637,7 +637,7 @@ describe(`HTTP Sync`, () => {
     }
 
-    const issueStream = new ShapeStream({
+    const issueStream = new ShapeStream<IssueRow>({
```
Since we added the `IssueRow` type argument here, i think we can remove it from L651 as `forEachMessage` should be able to infer it from the `issueStream` argument that is passed to it.

kevin-dp commented on packages/react-hooks/src/react-hooks.tsx:
```diff
@@ -40,13 +42,15 @@ export function getShapeStream(options: ShapeStreamOptions): ShapeStream {
 }
 
-export function getShape(shapeStream: ShapeStream): Shape {
+export function getShape<T extends Row = Row>(
```
I don't think we want to provide a default type argument here because the type argument can be inferred from the `shapeStream` argument.
So we want that `getShape(myShapeStream)` uses the type from `myShapeStream` but because we define the default type argument to be `Row` it would use `Row` (i think).

msfstef replied:
I've added a type test for this scenario and the `Shape` infers the type from the `ShapeStream` - even if it has a default, since the argument is a `ShapeStream<T>` it will use that to infer.

kevin-dp replied:
That's nice, so inferred type takes preference over default type, perfect!

kevin-dp replied:
Still, it's always going to infer the type argument from the actual argument, so there's no need to provide a default, is there?

msfstef replied:
That's fair - I'll remove it

kevin-dp commented on packages/react-hooks/src/react-hooks.tsx:
```diff
@@ -55,33 +59,38 @@ export function getShape(shapeStream: ShapeStream): Shape {
 }
 
-function shapeSubscribe(shape: Shape, callback: () => void) {
+function shapeSubscribe<T extends Row = Row>(
```
Same here, i would avoid the default type argument as it can be inferred from the `shape` argument.

kevin-dp commented on packages/react-hooks/src/react-hooks.tsx:
```diff
@@ -55,33 +59,38 @@ export function getShape(shapeStream: ShapeStream): Shape {
 }
 
-function parseShapeData(shape: Shape): UseShapeResult {
+function parseShapeData<T extends Row = Row>(
```
same about default type arg

kevin-dp commented on packages/react-hooks/src/react-hooks.tsx:
```diff
@@ -95,16 +104,20 @@ function identity<T>(arg: T): T {
 
-export function useShape<Selection = UseShapeResult>({
+export function useShape<
+  Selection = UseShapeResult<Row>,
```
same here about default type args

msfstef replied:
I'm going to switch these two around as I want the `Selection` to be `UseShapeResult<SourceData>` by default - but I also want the `SourceData` to be defaulted to `Row` cause `useShape` does not have any parameters that will allow it to infer the type, it needs to be explicitly typed.

msfstef replied:
(ended up not switching them around cause it feels more ergonomic to be able to _only_ specify the selection type)

kevin-dp reviewed (approved):
Nice type improvements!
Left some minor comments, mainly the same comment but added it everywhere it needs that change.

msfstef commented:
@kevin-dp  addressed all your comments!

@KyleAMathews added typings to all examples (missed one reference which usees Dates, I'll have to think about that one separately as we currently don't parse dates into JS dates) - I've also added a type template in the TS client docs which I think should be enough?

KyleAMathews commented on docs/api/clients/typescript.md:
```diff
@@ -32,7 +32,7 @@ The client exports a `ShapeStream` class for getting updates to shapes on a row-
 
 // Passes subscribers rows as they're inserted, updated, or deleted
-const stream = new ShapeStream({
+const stream = new ShapeStream<RowType>({
```
I wonder if this will be clear enough to people? They might think it's missing a type import and not realize it's something they need to write themselves. Perhaps just leave it out for now and the examples are good enough documentation for now?

kevin-dp replied:
JavaScript devs don't need to worry about this. TypeScript devs should be familiar with generics, so shouldn't be a problem?

kevin-dp replied:
I agree with @KyleAMathews here that it isn't clear what `RowType` is/should be.
Perhaps we can give an example or explain it in the text.

KyleAMathews commented on examples/nextjs-example/app/page.tsx:
```diff
@@ -58,7 +58,7 @@ async function clearItems() {
 
 export default function Home() {
-  const { data: items } = useShape(itemShape()) as unknown as { data: Item[] }
+  const { data: items } = useShape<Item>(itemShape())
```
SOOO GOOOD!

KyleAMathews commented:
> added typings to all examples (missed one reference which usees Dates, I'll have to think about that one separately as we currently don't parse dates into JS dates)

Add a custom parse for the date?

msfstef commented:
@KyleAMathews re: dates - might pick it up in separate PR
